File: scripts/mastodon.js
const instance = 'https://mastodon.social';
const username = '112950868045593874';

async function fetchLatestPost() {
    const container = document.getElementById('mastodon-post');
    if (!container) return;

    try {
        const response = await fetch(`${instance}/api/v1/accounts/${username}/statuses?limit=3&exclude_replies=true&exclude_reblogs=true`);
        const data = await response.json();

        if (data.length > 0) {
            displayPosts(data);
        } else {
            container.innerHTML = '<p>No recent posts found.</p>';
        }
    } catch (error) {
        console.error('Error fetching Mastodon posts:', error);
        container.innerHTML = '<p>No recent posts found.</p>';
    }
}

function displayPosts(posts) {
    const container = document.getElementById('mastodon-post');
    if (!container) return;

    const items = posts.map(post => `
        <div class="mastodon-post-item">
            <div class="mastodon-post-content">${post.content}</div>
            <p class="mastodon-post-date"><small>${new Date(post.created_at).toLocaleString()}</small></p>
        </div>
    `).join('');

    container.innerHTML = `<h2>Recent posts</h2>${items}`;
}

File: ssh-tui/mastodon.go
package main

import (
	"encoding/json"
	"html"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
)

const (
	mastodonInstance = "https://mastodon.social"
	mastodonAccount  = "112950868045593874"
)

type MastodonPost struct {
	Text string
	Date string
	URL  string
}

type mastodonMsg struct {
	posts []MastodonPost
	err   error
}

var htmlTagRe = regexp.MustCompile(`<[^>]+>`)

func stripHTML(s string) string {
	s = htmlTagRe.ReplaceAllString(s, " ")
	s = html.UnescapeString(s)
	// Collapse whitespace
	fields := strings.Fields(s)
	return strings.Join(fields, " ")
}

func fetchMastodonCmd() tea.Cmd {
	return func() tea.Msg {
		url := mastodonInstance + "/api/v1/accounts/" + mastodonAccount +
			"/statuses?limit=3&exclude_replies=true&exclude_reblogs=true"

		client := &http.Client{Timeout: 8 * time.Second}
		resp, err := client.Get(url)
		if err != nil {
			return mastodonMsg{err: err}
		}
		defer resp.Body.Close()

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return mastodonMsg{err: err}
		}

		var raw []struct {
			Content   string `json:"content"`
			CreatedAt string `json:"created_at"`
			URL       string `json:"url"`
		}
		if err := json.Unmarshal(body, &raw); err != nil {
			return mastodonMsg{err: err}
		}

		posts := make([]MastodonPost, 0, len(raw))
		for _, r := range raw {
			t, err := time.Parse(time.RFC3339, r.CreatedAt)
			date := r.CreatedAt
			if err == nil {
				date = t.Format("2006-01-02")
			}
			posts = append(posts, MastodonPost{
				Text: stripHTML(r.Content),
				Date: date,
				URL:  r.URL,
			})
		}
		return mastodonMsg{posts: posts}
	}
}
